### PR TITLE
create update submodules action

### PR DESCRIPTION
**Pull request by stephansama**
**Checklist**

- [ ] Latest changes from `main` have been merged
- [ ] Conflicts have been resolved
- [ ] The branch is pointing to `main`
- [ ] ESLint hasn't reported any issues.
- [ ] All unit tests pass
- [ ] Any new actions have an accompanying test and example workflows

### CODE REVIEW

File: update-git-submodules/package.json [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+		"@actions/core": "catalog:",
+		"@actions/exec": "catalog:",
+		"@types/ini": "catalog:",
+		"ini": "catalog:",
+		"markdown-table": "catalog:",
+		"zod": "catalog:"
+	},
+	"engines": {
+		"node": ">=24"
+	}
```
[the remark this answers is not in the document]
Comment (stephansama) [reply update-git-submodules/package.json:16]: Going to leave this at `>=24` for consistency with the rest of the monorepo — every other action in this repo (`do-not-track-polyfill`, `multi-deployments`, `todo-to-pr-comment-issue`, `generate-md-action-inputs`) also declares `engines.node: ">=24"` because that's the workspace's development Node version (see `.node-version`).

`engines.node` constrains the install environment; the action's actual runtime is determined by `runs.using: node20` in `action.yml`. End users consume the action via `uses: stephansama/actions/update-git-submodules@v1` → the prebuilt `dist/index.mjs` runs on the runner's node20 regardless. Aligning all actions to `>=20` is a worthwhile mono-repo-wide cleanup, but not for this PR.

File: update-git-submodules/src/index.ts [hunk 2]
```diff
@@ -0,0 +1,357 @@
...
+	return [full, full.slice(0, 7)];
+}
+
+export async function getLatestTag(cwd: string): Promise<string | undefined> {
+	await exec.getExecOutput("git", ["fetch", "--tags", "--force"], {
+		cwd,
+		ignoreReturnCode: true,
+	});
+	return getPreviousTag(cwd);
+}
```
[the remark this answers is not in the document]
Comment (stephansama) [reply update-git-submodules/src/index.ts:203]: Switched `getLatestTag` to `git tag --sort=-v:refname` taking the first line — `describe` semantics stay in `getPreviousTag` where they're correct ("what tag does the current commit follow"). Now newly-fetched tags that aren't yet reachable from HEAD are visible. Test added covering the version-sort and the no-tags case.

File: update-git-submodules/src/index.ts [hunk 3]
```diff
@@ -0,0 +1,357 @@
...
+	return entries.map(([name, raw]) => {
+		const entry = SubmoduleEntrySchema.parse(raw);
+		return {
+			name,
+			path: entry.path,
+			remoteName: getRemoteName(entry.url),
+			url: entry.url,
+		};
+	});
+}
```
[the remark this answers is not in the document]
Comment (stephansama) [reply update-git-submodules/src/index.ts:291]: Swapped `ini.parse` + Zod entry schema for `git config -f <file> --list`. Submodule names with dots (e.g. `[submodule "libs.foo"]`) now round-trip correctly — the regex uses greedy `submodule\.(.+)\.(path|url)=(.*)` so the right break point is forced. Dropped `ini` and `@types/ini` from the catalog and the package as a knock-on. Regression test for dotted names added.

File: update-git-submodules/src/index.ts [hunk 4]
```diff
@@ -0,0 +1,357 @@
...
+			return {
+				...s,
+				latestCommitSha,
+				latestShortCommitSha,
+				latestTag: undefined,
+				updated: latestCommitSha !== s.previousCommitSha,
+			};
+		}),
+	);
+}
```
[the remark this answers is not in the document]
Comment (stephansama) [reply update-git-submodules/src/index.ts:378]: Converted `updateToLatestCommit` to a sequential `for…of` so concurrent `git submodule update --remote` calls can't collide on the parent's `index.lock`. Added a test asserting ordering (submodule a's commands complete before b's begin).

File: update-git-submodules/src/index.ts [hunk 5]
```diff
@@ -0,0 +1,357 @@
...
+			return {
+				...s,
+				latestCommitSha,
+				latestShortCommitSha,
+				latestTag,
+				updated: latestTag !== s.previousTag,
+			};
+		}),
+	);
+}
```
[the remark this answers is not in the document]
Comment (stephansama) [reply update-git-submodules/src/index.ts:422]: Same fix — `updateToLatestTag` is now sequential `for…of`. Also flipped the `parsed.map(enrichSubmodule)` loop in `run()` from `Promise.all` to sequential for the same reason (`enrichSubmodule` issues git ops too).

File: update-git-submodules/src/index.ts [hunk 6]
```diff
@@ -0,0 +1,357 @@
...
+export async function configureAuth(token: string): Promise<void> {
+	if (!token) return;
+	const basic = Buffer.from(`x-access-token:${token}`).toString("base64");
+	await exec.getExecOutput("git", [
+		"config",
+		"--local",
+		AUTH_HEADER_CONFIG_KEY,
+		`AUTHORIZATION: basic ${basic}`,
+	]);
+}
```
[the remark this answers is not in the document]
Comment (stephansama) [reply update-git-submodules/src/index.ts:125]: Switched `configureAuth`/`cleanupAuth` to `--global` — same pattern `actions/checkout` uses for `persist-credentials`. The header now applies regardless of `cwd`, so submodule-cwd ops inherit it. Runner ephemerality + the `finally { cleanupAuth }` block covers leakage. I considered the per-command `git -c X=Y …` alternative but it would require threading the token through every helper signature; switched the asserted args in the existing tests instead.

File: update-git-submodules/src/index.ts [hunk 7]
```diff
@@ -0,0 +1,357 @@
...
+	if (filter.length > 0) {
+		result = result.filter(
+			(s) => filter.includes(s.name) || filter.includes(s.path),
+		);
+	}
+	if (strategy === "tag") {
+		result = result.filter((s) => s.previousTag !== undefined);
+	}
+	return result;
+}
```
[the remark this answers is not in the document]
Comment (stephansama) [reply update-git-submodules/src/index.ts:155]: Removed the tag-strategy filter in `filterSubmodules`. The `previousTag === undefined` case was already handled cleanly downstream (`buildPrBody` falls back to `—`, `updateToLatestTag` checks `!latestTag`), so first-run transitions to tag mode now actually work. Replaced the "drops untagged" test with one asserting the opposite.

File: update-git-submodules/src/index.ts [hunk 8]
```diff
@@ -0,0 +1,357 @@
...
+	return stdout.trim() || undefined;
+}
+
+export function getRemoteName(repoUrl: string): string | undefined {
+	for (const pattern of GITHUB_URL_PATTERNS) {
+		const match = repoUrl.match(pattern);
+		if (match) return match[1];
+	}
+	return undefined;
+}
```
[the remark this answers is not in the document]
Comment (stephansama) [reply update-git-submodules/src/index.ts]: Added `getParentRemoteUrl()` (runs `git remote get-url origin` with ignoreReturnCode) and extended `getRemoteName(url, parentRemoteUrl?)` to resolve `./` and `../`-prefixed URLs against the parent. SSH form (`git@host:owner/repo`) is normalized to `ssh://git@host/owner/repo` first so `new URL()` parses it, and the base gets a trailing-slash so resolution treats `parent.git` as a directory (otherwise `../sibling.git` would resolve to the wrong place). Threaded the parent URL through `parseGitmodulesFile` and called `getParentRemoteUrl()` from `run()`. Test cases added for both HTTPS and SSH parent URLs.

File: update-git-submodules/src/index.ts [hunk 9]
```diff
@@ -0,0 +1,389 @@
...
+export async function configureAuth(token: string): Promise<void> {
+	if (!token) return;
+	const basic = Buffer.from(`x-access-token:${token}`).toString("base64");
+	await exec.getExecOutput("git", [
+		"config",
+		"--global",
+		AUTH_HEADER_CONFIG_KEY,
+		`AUTHORIZATION: basic ${basic}`,
+	]);
+}
```
[the remark this answers is not in the document]
Comment (stephansama) [reply update-git-submodules/src/index.ts:125]: Addressed. `configureAuth` now also adds two `insteadOf` rewrites (`git@github.com:` → `https://github.com/` and `ssh://git@github.com/` → `https://github.com/`) so SSH-configured submodules get rewritten at git-config layer and pick up the extraheader. Tests added asserting both rewrites are emitted.

File: update-git-submodules/src/index.ts [hunk 10]
```diff
@@ -0,0 +1,389 @@
...
+		ignoreReturnCode: true,
+	});
+	const { exitCode, stdout } = await exec.getExecOutput(
+		"git",
+		["tag", "--sort=-v:refname"],
+		{ cwd, ignoreReturnCode: true },
+	);
+	if (exitCode !== 0) return undefined;
+	return stdout.split("\n")[0]?.trim() || undefined;
+}
```
[the remark this answers is not in the document]
Comment (stephansama) [reply update-git-submodules/src/index.ts:191]: Addressed. Switched `getLatestTag` to `git describe --tags --abbrev=0 FETCH_HEAD` — describe respects reachability so unrelated-branch tags don't surface, and FETCH_HEAD (not HEAD) means newly-fetched tags ARE visible (the bug from the previous round). `git tag --sort=-v:refname` stays as a fallback for the edge case where FETCH_HEAD isn't set or describe can't resolve. Tests updated to cover both the reachability case and the fallback case.

File: update-git-submodules/src/index.ts [hunk 11]
```diff
@@ -0,0 +1,389 @@
...
+
+export async function cleanupAuth(token: string): Promise<void> {
+	if (!token) return;
+	await exec.getExecOutput(
+		"git",
+		["config", "--global", "--unset-all", AUTH_HEADER_CONFIG_KEY],
+		{ ignoreReturnCode: true },
+	);
+}
```
[the remark this answers is not in the document]
Comment (stephansama) [reply update-git-submodules/src/index.ts:99]: Addressed alongside the `insteadOf` add — `cleanupAuth` now also runs `git config --global --unset-all url.https://github.com/.insteadOf` so self-hosted runners don't keep the rewrite. Test asserts both unset calls.

File: .github/workflows/examples/example-update-git-submodules.yml [hunk 12]
```diff
@@ -0,0 +1,15 @@
...
+  workflow_dispatch:
+jobs:
+  example:
+    runs-on: ubuntu-latest
+    permissions:
+      contents: read
+    steps:
+      - uses: actions/checkout@v4
+        with:
+          submodules: recursive
```
[the remark this answers is not in the document]
Comment (stephansama) [reply .github/workflows/examples/example-update-git-submodules.yml:12]: Addressed — added `persist-credentials: false` to the example workflow's checkout step.

File: .github/workflows/test-update-git-submodules.yml [hunk 13]
```diff
@@ -0,0 +1,110 @@
...
+          test -n "$URL_OUTPUT"
+          # url and remoteName must be distinct outputs (regression for original bug)
+          test "$URL_OUTPUT" != "$REMOTE_NAME"
+  release:
+    runs-on: ubuntu-latest
+    if: ${{github.event.workflow_run.conclusion == 'success'}}
+    steps:
+      - uses: actions/checkout@v4
+        with:
+          submodules: recursive
```
[the remark this answers is not in the document]
Comment (stephansama) [reply .github/workflows/test-update-git-submodules.yml]: Resolved by deleting the release job entirely — see my reply on https://github.com/stephansama/actions/pull/91#discussion_r3367910837 for the rationale. With the release job gone, the persist-credentials concern on that step is moot.

File: .github/workflows/test-update-git-submodules.yml [hunk 14]
```diff
@@ -0,0 +1,110 @@
...
+on:
+  pull_request:
+    branches: [main]
+    paths: ["update-git-submodules/**.ts"]
+  workflow_run:
+    workflows: ["release"]
+    types: ["completed"]
+  workflow_dispatch:
+env:
+  HUSKY: 0
```
[the remark this answers is not in the document]
Comment (stephansama) [reply .github/workflows/test-update-git-submodules.yml:11]: Addressed — added top-level `permissions: contents: read` and a `concurrency` block on `github.workflow + github.ref` with `cancel-in-progress: true`.

File: .github/workflows/test-update-git-submodules.yml [hunk 15]
```diff
@@ -0,0 +1,110 @@
...
+          # url and remoteName must be distinct outputs (regression for original bug)
+          test "$URL_OUTPUT" != "$REMOTE_NAME"
+  release:
+    runs-on: ubuntu-latest
+    if: ${{github.event.workflow_run.conclusion == 'success'}}
+    steps:
+      - uses: actions/checkout@v4
+        with:
+          submodules: recursive
+      - uses: stephansama/actions/update-git-submodules@v1
```
[the remark this answers is not in the document]
Comment (stephansama) [reply .github/workflows/test-update-git-submodules.yml]: Removed the release job entirely. It was a copy-paste from the action-template and didn't fit this action: running `stephansama/actions/update-git-submodules@v1` against the monorepo (which has no `.gitmodules`) would error out because `git config -f .gitmodules --list` exits non-zero. The preview job already exercises full end-to-end behavior against a real fixture (parent+child repos, BASELINE→ADVANCED transition, output assertions), and the new unit-level entry-point smoke test (`entry-point smoke test > module loads without TDZ when invoked directly`) covers "dist loads cleanly" after build. Adding fake output validation to the release job wouldn't have added value.

File: .github/workflows/test-update-git-submodules.yml [hunk 16]
```diff
@@ -0,0 +1,110 @@
...
+      - uses: actions/checkout@v4
+      - uses: pnpm/action-setup@v4
+      - uses: actions/setup-node@v4
+        with:
+          cache: "pnpm"
+          node-version-file: ".node-version"
+      - name: 📦 Installing dependencies
+        run: pnpm install
+      - name: 🏗️ Build all actions
+        run: pnpm run build
```
[the remark this answers is not in the document]
Comment (stephansama) [reply .github/workflows/test-update-git-submodules.yml:32]: Addressed — `actions/checkout@v4` in the preview job now has `persist-credentials: false`.

File: .github/workflows/test-update-git-submodules.yml [hunk 17]
```diff
@@ -0,0 +1,110 @@
...
+          echo "advanced=$ADVANCED_SHA" >> "$GITHUB_OUTPUT"
+      - name: 🚀 Run update-git-submodules against the fixture
+        id: update
+        working-directory: ${{ steps.fixture.outputs.parent }}
+        env:
+          INPUT_GITMODULESPATH: .gitmodules
+          INPUT_STRATEGY: commit
+          INPUT_SUBMODULES: ""
+          INPUT_TOKEN: ""
+        run: node ${{ github.workspace }}/update-git-submodules/dist/index.mjs
```
[the remark this answers is not in the document]
Comment (stephansama) [reply .github/workflows/test-update-git-submodules.yml:73]: Addressed — added `# No token needed — the fixture uses local file:// repositories.` above `INPUT_TOKEN: ""` in the preview-job env block.

File: .github/workflows/test-update-git-submodules.yml [hunk 18]
```diff
@@ -0,0 +1,110 @@
+name: 🧪 Test update-git-submodules
+on:
+  pull_request:
+    branches: [main]
+    paths: ["update-git-submodules/**.ts"]
+  workflow_run:
+    workflows: ["release"]
+    types: ["completed"]
```
[the remark this answers is not in the document]
Comment (stephansama) [reply .github/workflows/test-update-git-submodules.yml:8]: Not changing this — the generic `workflow_run` warning doesn't apply here because the workflow we depend on (`release.yml`) only triggers on `push` to `main` and `workflow_dispatch`. There's no `pull_request` trigger on `release`, so fork PRs cannot trigger it, so `workflow_run` here only ever fires from trusted main pushes. Switching to `workflow_call` would force the release workflow to know about this one, which is the wrong direction for the dependency.

File: update-git-submodules/package.json [hunk 19]
```diff
@@ -0,0 +1,17 @@
...
+		"build": "tsdown"
+	},
+	"devDependencies": {
+		"@actions/core": "catalog:",
+		"@actions/exec": "catalog:",
+		"markdown-table": "catalog:",
+		"zod": "catalog:"
+	},
+	"engines": {
+		"node": ">=24"
```
[the remark this answers is not in the document]
Comment (stephansama) [reply update-git-submodules/package.json:15]: Same point Gemini raised in the first round — see my earlier reply at https://github.com/stephansama/actions/pull/91#discussion_r3367857650 for the monorepo-consistency rationale. `engines.node` constrains the install environment, not the action runtime (`runs.using: node20` in action.yml governs that). Every other action in this monorepo declares `>=24`; changing only this one introduces inconsistency. Aligning the whole monorepo is a worthwhile separate change.

File: update-git-submodules/src/index.test.ts [hunk 20]
```diff
@@ -0,0 +1,687 @@
...
+	it("filters by path", async () => {
+		const { filterSubmodules } = await import("./index.js");
+		expect(filterSubmodules(enriched, ["vendor/b"])).toEqual([enriched[1]]);
+	});
+
+	it("keeps untagged submodules so tag strategy can transition them on first run", async () => {
+		const { filterSubmodules } = await import("./index.js");
+		expect(filterSubmodules(enriched, [])).toEqual(enriched);
+	});
+});
```
[the remark this answers is not in the document]
Comment (stephansama) [reply update-git-submodules/src/index.test.ts:505]: Addressed — added a test `filterSubmodules accepts mixed name and path filters in one call` asserting `filterSubmodules(enriched, ["a", "vendor/b"])` returns both entries.

File: update-git-submodules/src/index.test.ts [hunk 21]
```diff
@@ -0,0 +1,687 @@
...
+			"submodule update --remote vendor/a",
+		);
+		const indexOfUpdateB = callOrder.indexOf(
+			"submodule update --remote vendor/b",
+		);
+		expect(indexOfUpdateA).toBeLessThan(indexOfUpdateB);
+		expect(result[0]?.latestCommitSha).toBe("a".repeat(40));
+		expect(result[1]?.latestCommitSha).toBe("b".repeat(40));
+	});
+});
```
[the remark this answers is not in the document]
Comment (stephansama) [reply update-git-submodules/src/index.test.ts:743]: Addressed — added `updateToLatestCommit > propagates a git failure and stops the sequence` which makes `git submodule update --remote vendor/b` reject and asserts the whole function rejects with the right error. There's no partial-success contract by design (callers handle failure via the `run()` wrapper's setFailed).
